lib/fs/fakefs_test.go

```diff
@@ -172,3 +178,722 @@ func TestFakeFSRead(t *testing.T) {
+		if _, err := fs.Create(testCase.create); err != nil {
+			t.Fatal(err)
+		}
+
+		fd, err := fs.Open(testCase.open)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if got := fd.Name(); got != testCase.open {
+			t.Errorf("want %s, got %s", testCase.open, got)
+		}
+	}
+}
+
+func testFakeFSFileNameInsens(t *testing.T, fs Filesystem) {
+	var testCases = []struct {
+		create string
+		open   string
+	}{
+		{"BaZ", "bAz"},
+	}
+
+	for _, testCase := range testCases {
+		fd, err := fs.Create(testCase.create)
+		if err != nil {
+			t.Fatal(err)
+		}
+		fd.Close()
+
+		fd, err = fs.Open(testCase.open)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		defer fd.Close()
+
+		if got := fd.Name(); got != testCase.open {
+			t.Errorf("want %s, got %s", testCase.open, got)
+		}
+	}
+}
+
+func TestFakeFSRename(t *testing.T) {
+	fs := newFakeFilesystem("/qux")
+	if err := fs.MkdirAll("/foo/bar/baz", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/foo/bar/baz/qux")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	if err := fs.Rename("/foo/bar/baz/qux", "/foo/baz/bar/qux"); err == nil {
+		t.Errorf("rename to non-existent dir gave no error")
+	}
+
+	if err := fs.MkdirAll("/baz/bar/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Rename("/foo/bar/baz/qux", "/baz/bar/foo/qux"); err != nil {
+		t.Fatal(err)
+	}
+
+	var dirs = []struct {
+		dir   string
+		files []string
+	}{
+		{dir: "/", files: []string{"foo", "baz"}},
+		{dir: "/foo", files: []string{"bar"}},
+		{dir: "/foo/bar/baz", files: []string{}},
+		{dir: "/baz/bar/foo", files: []string{"qux"}},
+	}
+
+	for _, dir := range dirs {
+		assertDir(t, fs, dir.dir, dir.files)
+	}
+
+	if err := fs.Rename("/baz/bar/foo", "/baz/bar/FOO"); err != nil {
+		t.Fatal(err)
+	}
+
+	assertDir(t, fs, "/baz/bar", []string{"FOO"})
+	assertDir(t, fs, "/baz/bar/FOO", []string{"qux"})
+}
+
+func testFakeFSRenameInsensitive(t *testing.T, fs Filesystem) {
+	if err := fs.MkdirAll("/baz/bar/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.MkdirAll("/foO/baR/baZ", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/BAZ/BAR/FOO/QUX")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd.Close()
+
+	if err := fs.Rename("/Baz/bAr/foO/QuX", "/Foo/Bar/Baz/qUUx"); err != nil {
+		t.Fatal(err)
+	}
+
+	var dirs = []struct {
+		dir   string
+		files []string
+	}{
+		{dir: "/", files: []string{"foO", "baz"}},
+		{dir: "/foo", files: []string{"baR"}},
+		{dir: "/foo/bar/baz", files: []string{"qUUx"}},
+		{dir: "/baz/bar/foo", files: []string{}},
+	}
+
+	for _, dir := range dirs {
+		assertDir(t, fs, dir.dir, dir.files)
+	}
+
+	// the next rename can be done on Windows, and maybe elsewhere, but not on OS X
+	// so we're shooting for the lowest common denominator
+	if _, ok := fs.(*fakefs); ok || runtime.GOOS == "darwin" {
+		if err := fs.Rename("/foo/bar/BAZ", "/FOO/BAR/bAz"); err == nil {
+			t.Errorf("In-place case-only directory renames fail on OS X, should fail here, too: %s", err)
```

Comment:
I guess this is not an expected failure, so should be "not expected to fail here"?

Reply:
Should we have a mode to emulate this in fakefs?

Reply:
I'd actually love it if other maintainers would say something about this particular issue. Thing is: in-place case-only directory rename fails on OS X and doesn't fail on Windows. Judging by other tests, this is actually the only place Win and OSX behaviors differ.

My hunch is we have fakefs simulate OS X in this case, i.e. fail, as this is the lowest common. However, I can see how it would be beneficial to have both cases covered. Question is: do we really need two fakefs insensitive modes that would only differ in this particular behavior?

If we don't, then this error message is correct and what's expected.

Reply:
This is fascinating and I had no idea. Frankly I think this is a Go standard library bug and nothing else. The `rename` **system call** on macOS succeeds perfectly fine in `rename("test", "TEST")`, as shown by tracing `mv` on my system. However, *Go* just `stat`s both the old and new name and then returns "already exists" without even trying the rename.

Reply:
Raised it as https://github.com/golang/go/issues/35222

Yet it seems this should only be an issue for directories, and not files 🤷‍♂

Reply:
And only on `darwin`, actually. On windows Go behaves and doesn't throw this error.